webapp/src/Service/DOMJudgeService.php

```diff
@@ -1499,4 +1509,74 @@ public function getVerdicts(bool $mergeExternal = false): array
 
         return $verdicts;
     }
+
+    public function getScoreboardZip(
+        Request $request,
+        RequestStack $requestStack,
+        ?Contest $contest,
+        ScoreboardService $scoreboardService,
+        bool $forceUnfrozen = false
+    ) {
+        $data = $scoreboardService->getScoreboardTwigData(
+                request: $request,
+                response: null,
+                refreshUrl: '',
+                jury: false,
+                public: true,
+                static: true,
+                contest: $contest,
+                forceUnfrozen: $forceUnfrozen
+            ) + ['hide_menu' => true, 'current_contest' => $contest];
+
+        $request = $requestStack->pop();
+        // Use reflection to change the basepath property of the request, so we can detect
+        // all requested and assets
+        $requestReflection = new ReflectionClass($request);
+        $basePathProperty  = $requestReflection->getProperty('basePath');
+        $basePathProperty->setAccessible(true);
+        $basePathProperty->setValue($request, '/CHANGE_ME');
+        $requestStack->push($request);
+
+        $contestPage = $this->twig->render('public/scoreboard.html.twig', $data);
+
+        // Now get all assets that are used
+        $assetRegex = '|/CHANGE_ME/([/a-z0-9_\-\.]*)(\??[/a-z0-9_\-\.=]*)|i';
+        preg_match_all($assetRegex, $contestPage, $assetMatches);
+        $contestPage = preg_replace($assetRegex, '$1$2', $contestPage);
+
+        $zip = new ZipArchive();
+        if (!($tempFilename = tempnam($this->getDomjudgeTmpDir(), "contest-"))) {
+            throw new ServiceUnavailableHttpException(null, 'Could not create temporary file.');
+        }
+
+        $res = $zip->open($tempFilename, ZipArchive::OVERWRITE);
+        if ($res !== true) {
+            throw new ServiceUnavailableHttpException(null, 'Could not create temporary zip file.');
+        }
+        $zip->addFromString('index.html', $contestPage);
+
+        $publicPath = realpath(sprintf('%s/public/', $this->projectDir));
+        foreach ($assetMatches[1] as $file) {
+            $filepath = realpath($publicPath . '/' . $file);
+            if (!str_starts_with($filepath, $publicPath) &&
+                !str_starts_with($filepath, $this->vendorDir)
+            ) {
+                // Path outside of known good dirs: path traversal
+                continue;
+            }
+
+            $zip->addFile($filepath, $file);
+        }
+
+        // Also copy in the webfonts
+        $webfontsPath = $publicPath . '/webfonts/';
+        foreach (glob($webfontsPath . '*') as $fontFile) {
+            $fontName = basename($fontFile);
+            $zip->addFile($fontFile, 'webfonts/' . $fontName);
+        }
+        $zip->close();
+
+        return Utils::streamZipFile($tempFilename, 'contest.zip');
```

Comment:
It returns a different filename than the above urls suggest it will.

Reply:
I’m fine with renaming the url yes